cylc/flow/graph_parser.py

```diff
@@ -85,98 +85,110 @@ class GraphParser:
+        QUAL_FAM_SUBMIT_FAIL_ALL: (TASK_OUTPUT_SUBMIT_FAILED, False),
+        # Optional outputs.
+        QUAL_FAM_SUBMIT_FAIL_ANY: (TASK_OUTPUT_SUBMITTED, True),
+        QUAL_FAM_SUCCEED_ANY: (TASK_OUTPUT_SUCCEEDED, True),
+        QUAL_FAM_FAIL_ANY: (TASK_OUTPUT_SUCCEEDED, True),
+        QUAL_FAM_FINISH_ALL: (TASK_OUTPUT_SUCCEEDED, True),
+        QUAL_FAM_FINISH_ANY: (TASK_OUTPUT_SUCCEEDED, True),
+        QUAL_FAM_SUBMIT_ANY: (TASK_OUTPUT_SUBMITTED, True),
     }
 
     _RE_SUICIDE = r'(?:!)?'
     _RE_NODE = _RE_SUICIDE + TaskID.NAME_RE
-    _RE_NODE_OR_ACTION = r'(?:[!@])?' + TaskID.NAME_RE
+    _RE_NODE_OR_XTRIG = r'(?:[!@])?' + TaskID.NAME_RE
     _RE_PARAMS = r'<[\w,=\-+]+>'
     _RE_OFFSET = r'\[[\w\-\+\^:]+\]'
-    _RE_TRIG = TRIGGER + r'[\w\-]+'  # task or fam trigger
+    RE_QUAL = QUALIFIER + r'[\w\-]+'  # task or fam trigger
     _RE_OPT = r'\??'  # optional output indicator
 
     # Match if there are any spaces which could lead to graph problems
     REC_GRAPH_BAD_SPACES_LINE = re.compile(
         TaskID.NAME_RE +
-        r'''
+        rf"""
         (?<![\-+](?=\s*[0-9]))  # allow spaces after -+ if numbers follow
         (?!\s*[\-+]\s*[0-9])    # allow spaces before/after -+ if nums follow
         \s+                     # do not allow 'task<space>task'
-        ''' + TaskID.NAME_SUFFIX_RE, re.X)
+        {TaskID.NAME_SUFFIX_RE}
+        """,
+        re.X
+    )
 
-    # Match @actions (xtriggers, and more to come?).
-    REC_ACTION = re.compile(r'@[\w\-+%]+')
+    # Match @xtriggers.
+    REC_XTRIG = re.compile(r'@[\w\-+%]+')
 
     # Match fully qualified parameterized single nodes.
     REC_NODE_FULL = re.compile(
-        _RE_SUICIDE +
-        r'''
-        (?:(?:''' +
-        TaskID.NAME_RE + r'(?:' + _RE_PARAMS + r')?|' + _RE_PARAMS +
-        ''')                          # node name
-        )+                            # allow task<param> to repeat
-        (?:''' + _RE_OFFSET + r''')?  # optional cycle point offset
-        (?:''' + _RE_TRIG + r''')?    # optional trigger
-        (?:''' + _RE_OPT + r''')?     # optional optional indicator
-        ''', re.X)                    # end of string
-
-    # Extract node or action from left-side expressions after param expansion.
-    REC_NODES = re.compile(r'''
-        (''' + _RE_NODE_OR_ACTION + r''')  # node name
-        (''' + _RE_OFFSET + r''')?         # optional cycle point offset
-        (''' + _RE_TRIG + r''')?           # optional trigger
-        (''' + _RE_OPT + r''')?            # optional optional indicator
-    ''', re.X)
+        rf"""
+        {_RE_SUICIDE}
+        (?:(?:{TaskID.NAME_RE}             # node name
+        (?:{_RE_PARAMS})?|{_RE_PARAMS}))+  # allow task<param> to repeat
+        (?:{_RE_OFFSET})?                  # optional cycle point offset
+        (?:{RE_QUAL})?                     # optional trigger
+        (?:{_RE_OPT})?                     # optional optional indicator
```

Comment:
Actually, I don't think it's necessary to state whether the regex group is optional, as that should be reasonably clear from the `?` at the end
```suggestion
        (?:{_RE_OFFSET})?                  # cycle point offset
        (?:{RE_QUAL})?                     # qualifier
        (?:{_RE_OPT})?                     # optional output indicator
```

Reply:
Good point. I'll make those changes (forgot to change the comments yesterday)